api/python/t4/data_transfer.py

```diff
@@ -607,3 +610,120 @@ def _process_url(src):
+        'CSV': ['NONE', 'BZIP2', 'GZIP'],
+        }
+    # largely 1:1, but we still need to check that they're accepted formats.
+    ext_formats = {
+        '.parquet': 'parquet',
+        '.json': 'json',
+        '.jsonl': 'jsonl',
+        '.csv': 'csv',
+        '.tsv': 'tsv',
+        }
+    valid_s3_select_formats = {
+        'parquet': 'Parquet',
+        'json': 'JSON',
+        'jsonl': 'JSON',
+        'csv': 'CSV',
+        'tsv': 'CSV',
+        }
+
+    parsed_url = urlparse(url)
+
+    bucket, path, version_id = parse_s3_url(parsed_url)
+
+    # TODO: Use formats lib for this stuff
+    # use metadata to get format and compression
+    compression = None
+    format = meta.get('target')
+    if format is None:
+        format = meta.get('format', {}).get('name')
+        if format in ('bzip2', 'gzip'):
+            compression = format.upper()
+            format = meta.get('format', {}).get('contained_format', {}).get('name')
+
+    # use file extensions to get compression info, if none is present
+    exts = pathlib.Path(path).suffixes
+    if exts and not compression:
+        if exts[-1].lower() in accepted_compression:
+            compression = accepted_compression[exts.pop(-1)]
+    compression = compression if compression else 'NONE'
+
+    # use remaining file extensions to get format info, if none is present
+    if exts and not format:
+        ext = exts[-1].lower()
+        if ext in ext_formats:
+            if ext == '.tsv':
+                input_serialization = kwargs.setdefault('InputSerialization', {})
+                csv = input_serialization.setdefault('CSV', {})
+                if 'FieldDelimiter' not in csv:
+                    csv['FieldDelimiter'] = '\t'
+            format = ext_formats[ext]
+            s3_format = valid_s3_select_formats[format]
+            ok_compression = format_compression[s3_format]
+            if compression not in format_compression[s3_format]:
+                raise QuiltException("Compression {!r} not valid for select on format {!r}: "
+                                     "Expected {!r}".format(compression, s3_format, ok_compression))
+    if not format:
+        raise QuiltException("Unable to discover format for select on {!r}".format(url))
+
+    s3_format = valid_s3_select_formats[format]
+    input_serialization = kwargs.setdefault('InputSerialization', {})
+
+    if s3_format == 'JSON':
+        jsn = input_serialization.setdefault('JSON', {})
+        jsn['TYPE'] = "LINES" if format == 'jsonl' else "DOCUMENT"
+    elif s3_format == 'CSV':
+        csv = input_serialization.setdefault('CSV', {})
+        if format == 'tsv':
+            csv['FieldDelimiter'] = '\t'
+    elif s3_format == "Parquet":
+        input_serialization.setdefault('Parquet', {})
+
+    s3 = s3client if s3client else s3_client  # use passed in client, otherwise module-level client
+    if 'OutputSerialization' not in kwargs:
+        kwargs['OutputSerialization'] = {'JSON': {}}
```

Comment:
Is there a reason to hide this inside `kwargs` instead of explicitly putting it in the signature? Is this coming from a chained call or something?

Reply:
Direction on this?

Reply:
I tried to make it a bit more clear what exactly what was happening.  Processing the args is complex, but not at all a CPU intensive activity.  That being the case, I just process the args, then build them more explicitly, updating from kwargs and calling all in one area:

```
    select_kwargs = dict(
        Bucket=bucket,
        Key=path,
        Expression=query,
        ExpressionType=query_type,
        InputSerialization=input_serialization,
        OutputSerialization=output_serialization,
    )
    # Include user-specified passthrough options, overriding other options
    select_kwargs.update(kwargs)

    response = s3.select_object_content(**select_kwargs)
```

Reply:
lgtm, thanks for explaining